rafthttp/stream.go

```diff
@@ -0,0 +1,279 @@
+	// the data that starts with big-endian uint64 zero, it must return
+	// errHeartbeat.
+	decodeFrom(r io.Reader) (raftpb.Message, error)
+}
+
+type WriteFlushCloser interface {
+	io.WriteCloser
+	http.Flusher
+}
+
+type outgoingConn struct {
+	WriteFlushCloser
+	info http.Header
+}
+
+type streamWriter struct {
+	enc encoder
+
+	mu      sync.Mutex // guard field working
+	working bool
+	msgc    chan raftpb.Message
+	connc   chan outgoingConn
+	stopc   chan struct{}
+	done    chan struct{}
+}
+
+func startStreamWriter(enc encoder) *streamWriter {
+	w := &streamWriter{
+		enc:   enc,
+		msgc:  make(chan raftpb.Message, streamBufSize),
+		connc: make(chan outgoingConn),
+		stopc: make(chan struct{}),
+		done:  make(chan struct{}),
+	}
+	go w.run()
+	return w
+}
+
+func (cw *streamWriter) run() {
+	// msgc is available iff w is valid
+	var msgc chan raftpb.Message
+	var heartbeatc <-chan time.Time
+	var w WriteFlushCloser
+	tickc := time.Tick(ConnReadTimeout / 3)
+	reset := func(nw WriteFlushCloser) {
+		if w != nil {
+			w.Close()
+		}
+		if nw != nil {
+			heartbeatc, msgc = tickc, cw.msgc
+		} else {
+			heartbeatc, msgc = nil, nil
+		}
+		w = nw
+		cw.mu.Lock()
+		cw.working = nw != nil
+		cw.mu.Unlock()
+	}
+	for {
+		select {
+		case <-heartbeatc:
+			// send empty message as heartbeat
+			if err := binary.Write(w, binary.BigEndian, uint64(0)); err != nil {
+				log.Printf("rafthttp: encountered error heartbeating to connection: %v", err)
+				reset(nil)
+				continue
+			}
+			w.Flush()
+		case m := <-msgc:
+			if !cw.enc.isEllegible(m) {
+				log.Printf("rafthttp: silently dropped message %s because it is not ellegible", m.Type)
+				continue
+			}
+			if err := cw.enc.encodeTo(w, m); err != nil {
+				log.Printf("rafthttp: encountered error writing to connection: %v", err)
+				reset(nil)
+				continue
+			}
+			w.Flush()
+		case conn := <-cw.connc:
+			cw.enc.loadMetadata(conn.info)
+			reset(conn.WriteFlushCloser)
+		case <-cw.stopc:
+			if w != nil {
+				w.Close()
+			}
+			close(cw.done)
+			return
+		}
+	}
+}
+
+func (cw *streamWriter) canSend(m raftpb.Message) bool {
+	if !cw.enc.isEllegible(m) {
+		return false
+	}
+	cw.mu.Lock()
+	defer cw.mu.Unlock()
+	return cw.working
+}
+
+func (cw *streamWriter) setOutgoingConn(conn outgoingConn) bool {
+	select {
+	case cw.connc <- conn:
+		return true
+	case <-cw.done:
+		return false
+	}
+}
+
+func (cw *streamWriter) stop() {
+	close(cw.stopc)
+	<-cw.done
+}
+
+type streamDialer struct {
+	mu      sync.RWMutex // guard field u
+	u       string
+	tr      http.RoundTripper
+	from    types.ID
+	to      types.ID
+	cid     types.ID
+	typeStr string
+}
+
+func newStreamDialer(u string, tr http.RoundTripper, from, to, cid types.ID, typeStr string) *streamDialer {
+	return &streamDialer{
+		u:       u,
+		tr:      tr,
+		from:    from,
+		to:      to,
+		cid:     cid,
+		typeStr: typeStr,
+	}
+}
+
+func (d *streamDialer) dial(dec decoder) (io.ReadCloser, error) {
```

Comment:
i call it this way because it dials to a stream endpoint in the way to build a stream.